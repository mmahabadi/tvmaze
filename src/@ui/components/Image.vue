<script setup lang="ts">
import { defineProps } from 'vue'

defineProps<{ src?: string; alt?: string }>()
</script>
<template>
  <img v-if="src" :src="src" :alt="alt" class="w-full h-full object-cover" />
  <img
    v-else
    class="w-full h-full object-cover"
    src="@/assets/image-placeholder.jpg"
    alt="image-placeholder"
  />
</template>
