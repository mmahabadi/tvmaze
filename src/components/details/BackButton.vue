<script setup lang="ts">
import { useRouter } from 'vue-router'
import BackIcon from '@/@ui/components/icons/BackIcon.vue'
const router = useRouter()

const goBack = () => {
  router.go(-1)
}
</script>
<template>
  <button @click="goBack" class="btn-back">
    <div class="text">
      Back
      <BackIcon class="icon" />
    </div>
  </button>
</template>
<style>
.btn-back {
  @apply lg:w-1/4 hidden  w-full xl:flex justify-center items-center hover:bg-transparent border-2 transition duration-300 md:h-64 h-20 rounded font-medium;

  background-color: var(--primary-color);
  border-color: var(--primary-color);

  .text {
    @apply flex justify-center items-center gap-6 text-xl uppercase tracking-widest lg:rotate-90;
  }
  .icon {
    @apply h-9 w-9 md:-rotate-90;
  }
}
</style>
