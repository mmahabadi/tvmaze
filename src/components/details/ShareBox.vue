<script setup lang="ts">
import { defineProps } from 'vue'
import ShareIcon from '@/@ui/components/icons/ShareIcon.vue'
import IconPlay from '@/@ui/components/icons/IconPlay.vue'

defineProps<{
  language: string
}>()
</script>
<template>
  <div class="box">
    <div class="col1">
      <button class="share-btn">
        <ShareIcon />
      </button>
    </div>
    <div class="col2">
      <p>
        Language : <span class="ml-2 truncate">{{ language }}</span>
      </p>
    </div>
    <div class="col3">
      <button class="play-btn">
        <IconPlay />
        Watch
      </button>
    </div>
  </div>
</template>
<style scoped>
.box {
  @apply grid sm:grid-cols-5 grid-cols-3 gap-4 p-6 items-center bg-gray-900 border border-gray-800 rounded-lg;

  .col1 {
    @apply col-span-1 border-r hidden sm:flex;
  }
  .col2 {
    @apply col-span-2 flex-col hidden sm:flex font-medium text-sm;
  }
  .col3 {
    @apply sm:col-span-2 col-span-3 flex justify-end font-medium text-sm;
  }

  .share-btn {
    @apply w-10 h-10 flex justify-center items-center rounded-lg bg-white bg-opacity-20;
  }
  .play-btn {
    @apply py-4 transition border-2 rounded-full flex justify-center items-center gap-4 w-full sm:py-3;
    border-color: var(--primary-color);
  }
  .share-btn:hover,
  .play-btn:hover {
    background-color: var(--primary-color);
  }
}
</style>
