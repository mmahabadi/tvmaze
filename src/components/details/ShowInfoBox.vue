<script setup lang="ts">
import { defineProps } from 'vue'
import CalendarIcon from '@/@ui/components/icons/CalendarIcon.vue'
import ClockIcon from '@/@ui/components/icons/ClockIcon.vue'
import { Movie } from '@/types'

type propType = Partial<Pick<Movie, 'status' | 'year' | 'days' | 'time'>> & { genres?: string }

defineProps<propType>()
</script>
<template>
  <div class="box">
    <div class="status">{{ status }}</div>
    <div class="item">
      <span class="text">{{ genres }}</span>
    </div>
    <div class="item">
      <CalendarIcon class="text-primary" />
      <span class="text">
        {{ year }}
      </span>
    </div>
    <div class="item">
      <ClockIcon class="text-primary" />
      <span class="text">{{ days }} {{ time }}</span>
    </div>
  </div>
</template>
<style>
.box {
  @apply flex items-center gap-4 font-medium;

  .status {
    @apply text-xs px-2 py-1;
    background-color: var(--primary-color);
  }

  .item {
    @apply flex items-center gap-2;

    .text-primary {
      color: var(--primary-color);
    }
    .text {
      @apply text-xs font-medium;
    }
  }
}
</style>
