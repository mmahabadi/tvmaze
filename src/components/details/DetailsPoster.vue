<script setup lang="ts">
import DOMPurify from 'dompurify'
import { computed, defineProps } from 'vue'
import type { Movie } from '@/types'
import Image from '@/@ui/components/Image.vue'
import BackButton from './BackButton.vue'
import ShowInfoBox from './ShowInfoBox.vue'
import ShareBox from './ShareBox.vue'

const props = defineProps<{ item: Movie }>()
const { item } = props

const year = computed(() => {
  return item.ended?.split('-')[0]
})

const sanitizedSummary = computed(() => DOMPurify.sanitize(item.summary))
</script>
<template>
  <div data-testid="poster" class="poster-container">
    <Image :src="item.image?.original" :alt="item.name" class="large-image" />
    <div class="overlay">
      <div class="container">
        <div data-testid="thumbnail" class="image-content">
          <Image :src="item.image?.medium" :alt="item.name" class="thumbnail" />
        </div>
        <div class="main-content">
          <div class="inner-container">
            <h1 class="title">
              {{ item.name }}
            </h1>
            <ShowInfoBox
              :genres="item.genres.join(',')"
              :status="item.status"
              :days="item.schedule.days?.join(',')"
              :year="year"
              :time="item.schedule.time"
            />
            <p class="summary" v-html="sanitizedSummary"></p>
            <ShareBox :language="item.language" />
          </div>
          <div class="back-container">
            <BackButton />
          </div>
        </div>
      </div>
    </div>
  </div>
</template>
<style scoped>
.poster-container {
  @apply w-full h-full xl:max-h-[44rem] xl:overflow-hidden relative text-white;

  .large-image {
    @apply w-full hidden xl:inline-block h-full object-cover object-top;
  }
  .overlay {
    @apply xl:bg-black/90 flex-col xl:bg-opacity-90 xl:absolute top-0 left-0 right-0 bottom-0;

    .container {
      @apply px-3 mx-auto 2xl:px-32 flex-col items-center justify-center lg:grid grid-cols-3 py-10 lg:py-20 gap-8;
    }

    .image-content {
      @apply col-span-3 md:col-span-1 mx-auto w-[12rem] md:block lg:w-full border border-gray-800 rounded-lg overflow-hidden;

      .thumbnail {
        @apply xl:w-full xl:h-full xl:object-cover;
      }
    }
  }

  .main-content {
    @apply col-span-2 lg:grid grid-cols-5 gap-4 items-center;

    .inner-container {
      @apply col-span-5 xl:col-span-3 flex flex-col gap-10;

      .title {
        @apply xl:text-4xl capitalize font-sans text-2xl font-bold;
      }
    }

    .summary {
      @apply text-sm leading-7;
    }
    .back-container {
      @apply col-span-2 md:mt-0 mt-2 flex justify-center items-center;
    }
  }
}
</style>
