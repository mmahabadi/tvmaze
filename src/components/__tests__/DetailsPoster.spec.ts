import { describe, it, expect, beforeEach, vi } from 'vitest'
import { mount, VueWrapper } from '@vue/test-utils'
import DetailsPoster from '../details/DetailsPoster.vue'
import Image from '@/@ui/components/Image.vue'
import BackButton from '../details/BackButton.vue'
import ShowInfoBox from '../details/ShowInfoBox.vue'
import ShareBox from '../details/ShareBox.vue'
import type { Movie } from '@/types'

describe('DetailsPoster', () => {
  let wrapper: VueWrapper<unknown>
  const props = {
    item: {
      id: 525,
      name: 'Gilmore Girls',
      language: 'English',
      genres: ['Drama', 'Comedy', 'Romance'],
      status: 'Ended',
      ended: '2007-05-15',
      schedule: {
        time: '21:00',
        days: ['Tuesday']
      },
      rating: {
        average: 8.1
      },
      image: {
        medium: 'https://static.tvmaze.com/uploads/images/medium_portrait/4/11308.jpg',
        original: 'https://static.tvmaze.com/uploads/images/original_untouched/4/11308.jpg'
      },
      summary:
        '<p><b>Gilmore Girls</b> is a drama centering around the relationship between a thirtysomething single mother and her teen daughter living in Stars Hollow, Connecticut.</p>'
    } as Movie
  }

  beforeEach(() => {
    wrapper = mount(DetailsPoster, {
      props,
      global: {
        stubs: {
          Image,
          ShowInfoBox,
          ShareBox,
          BackButton
        }
      }
    })
  })

  it('renders correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
})
