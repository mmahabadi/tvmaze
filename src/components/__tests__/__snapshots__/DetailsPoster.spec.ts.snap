// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DetailsPoster > renders correctly 1`] = `
"<div data-v-0f6e4b97="" class="poster-container"><img data-v-0f6e4b97="" src="https://static.tvmaze.com/uploads/images/original_untouched/4/11308.jpg" alt="Gilmore Girls" class="w-full h-full object-cover large-image">
  <div data-v-0f6e4b97="" class="overlay">
    <div data-v-0f6e4b97="" class="container">
      <div data-v-0f6e4b97="" class="image-content"><img data-v-0f6e4b97="" src="https://static.tvmaze.com/uploads/images/medium_portrait/4/11308.jpg" alt="Gilmore Girls" class="w-full h-full object-cover thumbnail"></div>
      <div data-v-0f6e4b97="" class="main-content">
        <div data-v-0f6e4b97="" class="inner-container">
          <h1 data-v-0f6e4b97="" class="title">Gilmore Girls</h1>
          <div data-v-0f6e4b97="" class="box">
            <div class="status">Ended</div>
            <div class="item"><span class="text">Drama,Comedy,Romance</span></div>
            <div class="item"><svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 448 512" class="text-subMain w-3 h-3 text-primary" height="1em" width="1em" xmlns="http://www.w3.org/2000/svg">
                <path d="M148 288h-40c-6.6 0-12-5.4-12-12v-40c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12zm108-12v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm96 0v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm-96 96v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm-96 0v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm192 0v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm96-260v352c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V112c0-26.5 21.5-48 48-48h48V12c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v52h128V12c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v52h48c26.5 0 48 21.5 48 48zm-48 346V160H48v298c0 3.3 2.7 6 6 6h340c3.3 0 6-2.7 6-6z"></path>
              </svg><span class="text">2007</span></div>
            <div class="item"><svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 24 24" class="text-subMain w-3 h-3 text-primary" height="1em" width="1em" xmlns="http://www.w3.org/2000/svg">
                <path d="M12 2C6.486 2 2 6.486 2 12s4.486 10 10 10 10-4.486 10-10S17.514 2 12 2zm0 18c-4.411 0-8-3.589-8-8s3.589-8 8-8 8 3.589 8 8-3.589 8-8 8z"></path>
                <path d="M13 7h-2v6h6v-2h-4z"></path>
              </svg><span class="text">Tuesday 21:00</span></div>
          </div>
          <p data-v-0f6e4b97="" class="summary">
          <p><b>Gilmore Girls</b> is a drama centering around the relationship between a thirtysomething single mother and her teen daughter living in Stars Hollow, Connecticut.</p>
          </p>
          <div data-v-b00b0726="" data-v-0f6e4b97="" class="box">
            <div data-v-b00b0726="" class="col1"><button data-v-b00b0726="" class="share-btn"><svg data-v-b00b0726="" stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 448 512" height="1em" width="1em" xmlns="http://www.w3.org/2000/svg">
                  <path d="M352 320c-22.608 0-43.387 7.819-59.79 20.895l-102.486-64.054a96.551 96.551 0 0 0 0-41.683l102.486-64.054C308.613 184.181 329.392 192 352 192c53.019 0 96-42.981 96-96S405.019 0 352 0s-96 42.981-96 96c0 7.158.79 14.13 2.276 20.841L155.79 180.895C139.387 167.819 118.608 160 96 160c-53.019 0-96 42.981-96 96s42.981 96 96 96c22.608 0 43.387-7.819 59.79-20.895l102.486 64.054A96.301 96.301 0 0 0 256 416c0 53.019 42.981 96 96 96s96-42.981 96-96-42.981-96-96-96z"></path>
                </svg></button></div>
            <div data-v-b00b0726="" class="col2">
              <p data-v-b00b0726=""> Language : <span data-v-b00b0726="" class="ml-2 truncate">English</span></p>
            </div>
            <div data-v-b00b0726="" class="col3"><button data-v-b00b0726="" class="play-btn"><svg data-v-b00b0726="" class="svg-icon" style="width: 1em; height: 1em; vertical-align: middle; fill: currentColor; overflow: hidden;" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg">
                  <path d="M852.727563 392.447107C956.997809 458.473635 956.941389 565.559517 852.727563 631.55032L281.888889 993.019655C177.618644 1059.046186 93.090909 1016.054114 93.090909 897.137364L93.090909 126.860063C93.090909 7.879206 177.675064-35.013033 281.888889 30.977769L852.727563 392.447107 852.727563 392.447107Z"></path>
                </svg> Watch </button></div>
          </div>
        </div>
        <div data-v-0f6e4b97="" class="back-container"><button data-v-0f6e4b97="" class="btn-back">
            <div class="text"> Back <svg xmlns="http://www.w3.org/2000/svg" shape-rendering="geometricPrecision" text-rendering="geometricPrecision" image-rendering="optimizeQuality" fill-rule="evenodd" clip-rule="evenodd" viewBox="0 0 512 512.001" class="icon">
                <path fill="#FFF" fill-rule="nonzero" d="M512 256c0 70.685-28.658 134.696-74.981 181.019-46.324 46.324-110.327 74.982-181.019 74.982-70.692 0-134.695-28.658-181.019-74.982C28.658 390.696 0 326.692 0 256c0-70.691 28.658-134.695 74.981-181.019C121.305 28.658 185.308 0 256 0c70.692 0 134.695 28.658 181.019 74.981C483.342 121.305 512 185.309 512 256zm-200.641-79.002c17.861-30.984-6.963-46.317-25.063-29.013L185.25 236.726c-16.136 16.13-16.136 22.427 0 38.563l101.046 88.742c18.187 17.18 42.867 2.202 25.063-29.021l-46.932-79.002 46.932-79.01z"></path>
              </svg></div>
          </button></div>
      </div>
    </div>
  </div>
</div>"
`;
